lib/gemstash/storage.rb

```diff
@@ -7,343 +7,65 @@
-    #
-    # @param key [Symbol] the key of the content to delete
-    # @return [Gemstash::Resource] self for chaining purposes
-    def delete(key)
-      return self unless exist?(key)
-
-      begin
-        File.delete(content_filename(key))
-      rescue StandardError => e
-        log_error "Failed to delete stored content at #{content_filename(key)}", e, level: :warn
-      end
-
-      begin
-        File.delete(properties_filename) unless content?
-      rescue StandardError => e
-        log_error "Failed to delete stored properties at #{properties_filename}", e, level: :warn
+    def self.for(name)
+      storage_service = gemstash_env.config[:storage_adapter]
+      case storage_service
+      when "local"
+        new(gemstash_env.base_file(name), storage_service)
+      when "s3"
+        new(File.join(gemstash_env.config[:s3_path], name), storage_service)
+      else
+        raise Gemstash::Storage::InvalidStorage, storage_service
       end
-
-      self
-    ensure
-      reset
-    end
-
-  private
-
-    def load(key)
-      raise "Resource #{@name} has no #{key.inspect} content to load" unless exist?(key)
-
-      load_properties # Ensures storage version is checked
-      @content ||= {}
-      @content[key] = read_file(content_filename(key))
-    end
-
-    def load_properties(force = false)
-      return if @properties && !force
-      return unless File.exist?(properties_filename)
-
-      @properties = YAML.load_file(properties_filename) || {}
-      check_resource_version
-    end
-
-    def check_resource_version
-      version = @properties[:gemstash_resource_version]
-      return if version <= Gemstash::Resource::VERSION
-
-      reset
-      raise Gemstash::Resource::VersionTooNew.new(name, folder, version)
-    end
-
-    def reset
-      @content = nil
-      @properties = nil
-    end
-
-    def content?
-      return false unless Dir.exist?(@folder)
-
-      entries = Dir.entries(@folder).reject {|file| file =~ /\A\.\.?\z/ || file == "properties.yaml" }
-      !entries.empty?
     end
 
-    def sanitize(name)
-      name.gsub(/[^a-zA-Z0-9_]/, "_")
-    end
-
-    def save_content(key, content)
-      store(content_filename(key), content)
-      @content ||= {}
-      @content[key] = content
-    end
-
-    def save_properties(props)
-      props ||= {}
-      props = { gemstash_resource_version: Gemstash::Resource::VERSION }.merge(props)
-      store(properties_filename, props.to_yaml)
-      @properties = props
-    end
-
-    def store(filename, content)
-      FileUtils.mkpath(@folder) unless Dir.exist?(@folder)
-      save_file(filename) { content }
-    end
-
-    def save_file(filename)
-      content = yield
-      gemstash_env.atomic_write(filename) {|f| f.write(content) }
-    end
-
-    def read_file(filename)
-      File.open(filename, "rb", &:read)
+    def self.metadata
+      storage_service = gemstash_env.config[:storage_adapter]
+      case storage_service
+      when "local"
```

Comment:
We are doing this check of "local" and "s3" in a few places, maybe we can put it in a method

Reply:
Ideally, this would be solved by #264, it should handle all storage_service selection choices in one dedicated place.